libnmstate/metadata.py

```diff
@@ -152,3 +159,23 @@ def _generate_link_master_metadata(ifaces_desired_state,
                         master_has_no_slaves_specified_in_desired):
                     set_metadata_func(
                         master_state, ifaces_desired_state[slave])
+
+
+def _generate_route_metadata(desired_state):
+    """
+    Save routes under interface IP protocol so that nm/ipv4.py or nm/ipv6.py
+    could include route configuration in `create_setting()`.
+    Currently route['next-hop-interface'] is mandatory.
+    """
+    for iface_name, routes in six.viewitems(desired_state.iface_routes):
+        iface_state = desired_state.interfaces.get(iface_name, {})
+        for family in (Interface.IPV4, Interface.IPV6):
+            if family in iface_state:
+                iface_state[family]['_routes'] = []
+            else:
+                iface_state[family] = {'_routes': []}
+        for route in routes:
+            if iplib.is_ipv6_address(route[Route.DESTINATION]):
```

Comment:
Suggestion:

```suggestion
            if iplib.is_ipv6_address(route[Route.DESTINATION]):
                family = Interface.IPV6
            else:
                family = Interface.IPV4
            iface_state[family].setdefault('_routes', []).append(route)
```

Then you can remove the code above to initialize the routes metadata with an empty list. Please note, that `_routes` here should be replace with the constant that should be introduced IMHO.

Reply:
Still need to handle `family not in iface_state` and calling `setdefault()` for every route seems time wasting.

Reply:
AFAIU #340 `family in iface_state` is always true now. Does `setdefault()` have a serious performance impact over `[ROUTES]`? The latter is also calling `__getitem__()`.